src/PerformanceSensitiveAnalyzers/Core/ExplicitAllocationAnalyzer.cs

```diff
@@ -0,0 +1,112 @@
+        private static readonly LocalizableString s_localizablAnonymousObjectCreationRuleTitleAndMessage = new LocalizableResourceString(nameof(PerformanceSensitiveAnalyzersResources.AnonymousNewObjectRuleTitleAndMessage), PerformanceSensitiveAnalyzersResources.ResourceManager, typeof(PerformanceSensitiveAnalyzersResources));
+        private static readonly LocalizableString s_localizableLetCauseRuleTitleAndMessage = new LocalizableResourceString(nameof(PerformanceSensitiveAnalyzersResources.LetCauseRuleTitleAndMessage), PerformanceSensitiveAnalyzersResources.ResourceManager, typeof(PerformanceSensitiveAnalyzersResources));
+
+        internal static DiagnosticDescriptor ArrayCreationRule = new DiagnosticDescriptor(
+            ArrayCreationRuleId,
+            s_localizableArrayCreationRuleTitleAndMessage,
+            s_localizableArrayCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        internal static DiagnosticDescriptor ObjectCreationRule = new DiagnosticDescriptor(
+            ObjectCreationRuleId,
+            s_localizableObjectCreationRuleTitleAndMessage,
+            s_localizableObjectCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        internal static DiagnosticDescriptor AnonymousObjectCreationRule = new DiagnosticDescriptor(
+            AnonymousObjectCreationRuleId,
+            s_localizablAnonymousObjectCreationRuleTitleAndMessage,
+            s_localizablAnonymousObjectCreationRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true,
+            helpLinkUri: "http://msdn.microsoft.com/en-us/library/bb397696.aspx");
+
+        internal static DiagnosticDescriptor LetCauseRule = new DiagnosticDescriptor(
+            LetCauseRuleId,
+            s_localizableLetCauseRuleTitleAndMessage,
+            s_localizableLetCauseRuleTitleAndMessage,
+            DiagnosticCategory.Performance,
+            DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        private static readonly object[] EmptyMessageArgs = Array.Empty<object>();
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            ArrayCreationRule,
+            ObjectCreationRule,
+            AnonymousObjectCreationRule,
+            LetCauseRule);
+
+        protected override ImmutableArray<OperationKind> Operations => ImmutableArray.Create(
+            OperationKind.ArrayCreation,
+            OperationKind.ObjectCreation,
```

Comment:
❗️  This needs `OperationKind.DelegateCreation`. The handling appears to be the following:

```
if (context.Operation is IDelegateCreationOperation delegateCreationOperation)
{
    if (!delegateCreationOperation.IsImplicit)
    {
        context.ReportDiagnostic(Diagnostic.Create(ObjectCreationRule, context.Operation.Syntax.GetLocation(), EmptyMessageArgs));
    }

    return;
}
```

Here are tests:

```
        [Fact]
        public async Task ExplicitAllocation_ExplicitDelegateCreation()
        {
            var sampleProgram =
@"using System;
using Roslyn.Utilities;

public class MyClass
{
    [PerformanceSensitive(""uri"")]
    public void Testing(object sender, EventArgs e)
    {
        var handler = new EventHandler(Testing);
    }
}";
            await VerifyCS.VerifyAnalyzerAsync(sampleProgram,
                VerifyCS.Diagnostic(ExplicitAllocationAnalyzer.ObjectCreationRule).WithLocation(9, 23));
        }

        [Fact]
        public async Task ExplicitAllocation_ImplicitDelegateCreation()
        {
            var sampleProgram =
@"using System;
using Roslyn.Utilities;

public class MyClass
{
    [PerformanceSensitive(""uri"")]
    public void Testing(object sender, EventArgs e)
    {
        EventHandler handler = Testing;
    }
}";
            await VerifyCS.VerifyAnalyzerAsync(sampleProgram);
        }
```

Reply:
This change results in a new diagnostic for code with the following form:

```
(Action)(Console.WriteLine)
```

I believe that is a reasonable change. However, we should file a follow-up bug to determine the best way to handle cases where the delegate will be cached:

```
(Action)(() => Console.WriteLine())
```